File: python/insertion_sort.py
from base_algorithm import BaseAlgorithm
from isorting_algorithm import ISortingAlgorithm

class InsertionSort(BaseAlgorithm, ISortingAlgorithm):
    def __init__(self):
        super().__init__()
        self.name = "Ordenación por Inserción"
        self.minimal_complexity = "O(n)"
        self.average_complexity = "O(n^2)"
        self.maximal_complexity = "O(n^2)"

    def sort(self, array):
        n = len(array)
        for i in range(1, n):
            key = array[i]
            j = i - 1
            while j >= 0 and array[j] > key:
                array[j + 1] = array[j]
                j -= 1
            array[j + 1] = key

    def get_elemental_operations(self, n):
        return n * n // 4

File: src/sorting/InsertionSort.java
package sorting;

import base.BaseAlgorithm;
import interfaces.ISortingAlgorithm;

public class InsertionSort extends BaseAlgorithm implements ISortingAlgorithm {

    public InsertionSort() {
        this.name = "Ordenación por Inserción";
        this.minimalComplexity = "O(n)";
        this.averageComplexity = "O(n²)";
        this.maximalCompleity = "O(n²)";
    }

    @Override
    public void sort(int[] array) {
        int n = array.length;

        for (int i = 0; i < n; i++) {
            int key = array[i];
            int j = i - 1;

            while (j >= 0 && array[j] > key) {
                array[j + 1] = array[j];
                j--;
            }
            array[j + 1] = key;
        }
    }

    @Override
    public Long getElementalOperations(int n) {
        return (long) n * n / 4;
    }
}
